### PR TITLE
feat: extend external signature endpoints to facilitate pre-authorized OID4VCI flow

### LINKED ISSUE #726: OID4VCI - New endpoints for client-provided signatures
**Describe the solution you'd like**
- extend the already existing optional feature in the wallet api for external signatures
- focus on pre-authorized code flow
- add a new endpoint:
  - Inputs: same as the regular `/useOfferRequest` endpoint
  - Outputs: proof of did possession JWT header and payload params and whatever else is needed to continue with the flow.
- add a new endpoint:
  - Inputs: signed and compact serialized proof of did possession and other other params returned by the previous endpoint.
  - Outputs: same as the regular `/useOfferRequest` endpoint
- make it stateless, i.e., do not store any information in the wallet's database in between these calls.

**Additional context**
Based on the work that is already in place for the respective OID4VP handling.

### PR DESCRIPTION
**Pull request by cpatsonakis**
## Description

Provide a clear and concise description of the changes made in this pull request.

Fixes #726

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist

- [x] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests
- [ ] Do we need to implement analytics?
- [ ] Will this be part of a product update? If yes, please write one phrase about this update.
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/service/exchange/ProofOfPossessionParameterFactory.kt [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import id.walt.oid4vc.data.CredentialOffer
+import id.walt.oid4vc.data.OfferedCredential
+import id.walt.oid4vc.data.ProofOfPossession
+import id.walt.oid4vc.data.ProofType
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonElement
+import kotlinx.serialization.json.encodeToJsonElement
+
+object ProofOfPossessionParameterFactory {
```
Comment (mikeplotean) [inline waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/service/exchange/ProofOfPossessionParameterFactory.kt:16]: nitpicking alert!
this could be refactored even more, have a separate method for building the proof-type from each key and did, for cwt and jwt, respectively and also have the instantiation of possesion-parameters extracted into separate methods, e.g.:
```kotlin
private fun cwtProofOfPossessionParameters(it: ProofOfPossession.CWTProofBuilder) =
    ProofOfPossessionParameters(
        ProofType.cwt,
        Json.encodeToJsonElement(it.headers.toCBOR()),
        Json.encodeToJsonElement(it.payload.toCBOR()),
    )
```
and for jwt:
```kotlin
private fun jwtProofOfPossessionParameters(it: ProofOfPossession.JWTProofBuilder) =
    ProofOfPossessionParameters(
        ProofType.jwt,
        it.headers.toJsonElement(),
        it.payload.toJsonElement(),
    )
```
then building the proof type for cwt from key:
```kotlin
private suspend fun cwtProofTypeFromKey(
    credentialOffer: CredentialOffer,
    nonce: String?,
    key: Key
) = let {
    ProofOfPossession.CWTProofBuilder(
        issuerUrl = credentialOffer.credentialIssuer,
        nonce = nonce,
        coseKey = OneKey(
            ECKey.parse(key.getPublicKey().exportJWK()).toECPublicKey(),
            null
        ).AsCBOR().EncodeToBytes(),
    ).let {
        cwtProofOfPossessionParameters(it)
    }
}
```
and from did:
```kotlin
private fun cwtProofTypeFromDid(credentialOffer: CredentialOffer, did: String, nonce: String?) = let {
    ProofOfPossession.CWTProofBuilder(
        issuerUrl = credentialOffer.credentialIssuer,
        clientId = did,
        nonce = nonce,
    ).let {
        cwtProofOfPossessionParameters(it)
    }
}
```
and the same thing for jwt:
```kotlin
private suspend fun jwtProofTypeFromKey(
    credentialOffer: CredentialOffer,
    nonce: String?,
    key: Key
) = let {
    ProofOfPossession.JWTProofBuilder(
        issuerUrl = credentialOffer.credentialIssuer,
        nonce = nonce,
        keyJwk = key.getPublicKey().exportJWKObject(),
        keyId = key.getKeyId(),
    ).let {
        jwtProofOfPossessionParameters(it)
    }
}
```
```kotlin
private fun jwtProofTypeFromDid(
    credentialOffer: CredentialOffer,
    did: String,
    nonce: String?,
    keyId: String
) = let {
    ProofOfPossession.JWTProofBuilder(
        issuerUrl = credentialOffer.credentialIssuer,
        clientId = did,
        nonce,
        keyId,
    ).let {
        jwtProofOfPossessionParameters(it)
    }
}
```

then `keyProofParameters` would look like:
```kotlin
private suspend fun keyProofParameters(
    did: String,
    offeredCredential: OfferedCredential,
    credentialOffer: CredentialOffer,
    nonce: String?
): ProofOfPossessionParameters {
    val key = DidService.resolveToKey(did).getOrThrow()
    val proofType = offeredCredential.proofTypesSupported?.keys?.first() ?: ProofType.jwt
    return when (proofType) {
        ProofType.cwt -> cwtProofTypeFromKey(credentialOffer, nonce, key)
        ProofType.ldp_vp -> TODO("ldp_vp proof not yet implemented")
        else -> jwtProofTypeFromKey(credentialOffer, nonce, key)
    }
}
```
and the `didProofParameters`:
```kotlin
private fun didProofParameters(
    did: String,
    keyId: String,
    offeredCredential: OfferedCredential,
    credentialOffer: CredentialOffer,
    nonce: String?
): ProofOfPossessionParameters {
    val proofType = offeredCredential.proofTypesSupported?.keys?.first() ?: ProofType.jwt
    return when (proofType) {
        ProofType.cwt -> cwtProofTypeFromDid(credentialOffer, did, nonce)
        ProofType.ldp_vp -> TODO("ldp_vp proof not yet implemented")
        else -> jwtProofTypeFromDid(credentialOffer, did, nonce, keyId)
    }
}
```

File: waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/web/controllers/exchange/OpenAPICommons.kt [hunk 2]
```diff
@@ -1,5 +1,6 @@
 package id.walt.webwallet.web.controllers.exchange
```
Comment (mikeplotean) [inline waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/web/controllers/exchange/OpenAPICommons.kt:1]: not part of this pr, but maybe worth renaming
- `rootPath` to smth. like `exchangeRootPath` 
- and `route` to `exchangeRoute`

File: waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/web/controllers/exchange/ExchangeExternalSignaturesController.kt [hunk 3]
```diff
@@ -355,6 +350,176 @@ fun Application.exchangeExternalSignatures() = walletRoute {
...
+            response(OpenAPICommons.useOfferRequestEndpointResponseParams())
+        }) {
+            val walletService = getWalletService()
+            val req = call.receive<SubmitOID4VCIRequest>()
+            logger.debug { "Request: $req" }
+
+            val did = req.did ?: walletService.listDids().firstOrNull()?.did
+            ?: throw IllegalArgumentException("No DID to use supplied and no DID was found in wallet.")
+
+            runCatching {
```
Comment (mikeplotean) [inline waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/web/controllers/exchange/ExchangeExternalSignaturesController.kt:500]: this could probably be simplified to smth. like:
```kotlin
runCatching {
    WalletServiceManager
        .externalSignatureClaimStrategy
        .submitCredentialClaim(
            tenantId = walletService.tenant,
            accountId = walletService.accountId,
            walletId = walletService.walletId,
            pending = req.requireUserInput ?: true,
            did = did,
            offerURL = req.offerURL,
            credentialIssuerURL = req.credentialIssuer,
            accessToken = req.accessToken,
            offeredCredentialProofsOfPossession = req.offeredCredentialProofsOfPossession,
        )
}.getOrElse { throw IllegalArgumentException("Unknown error") }.let {
    context.respond(HttpStatusCode.OK, it)
}
```
in case of an exception it will throw the IAE which gets handled by [StatusPages](https://github.com/walt-id/waltid-identity/blob/5ec0016b4174cb76e31fbb04343bf0843c9f0ea3/waltid-services/waltid-service-commons/src/main/kotlin/id/walt/commons/web/plugins/StatusPages.kt#L34) as a `BadRequest`

File: waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/web/controllers/exchange/ExchangeExternalSignaturesController.kt [hunk 4]
```diff
@@ -355,6 +350,176 @@ fun Application.exchangeExternalSignatures() = walletRoute {
...
+            val did = req.did ?: walletService.listDids().firstOrNull()?.did
+            ?: throw IllegalArgumentException("No DID to use supplied and no DID was found in wallet.")
+
+            //this can't fail due to the above block
+            val walletDID = DidsService.get(walletService.walletId, did)!!
+            logger.debug { "Retrieved wallet DID: $walletDID" }
+            val authKeyId = ExchangeUtils.getFirstAuthKeyIdFromDidDocument(walletDID.document)
+            logger.debug { "Resolved authorization keyId: $authKeyId" }
+
+            runCatching {
```
Comment (mikeplotean) [inline waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/web/controllers/exchange/ExchangeExternalSignaturesController.kt:424]: this could probably be simplified to smth. like:
```kotlin
runCatching {
    WalletServiceManager.externalSignatureClaimStrategy.prepareCredentialClaim(
        did = walletDID.did,
        keyId = authKeyId,
        offerURL = offer,
    )
}.getOrElse { throw IllegalStateException("") }.let { prepareClaimResult ->
    val responsePayload = PrepareOID4VCIResponse(
        did = walletDID.did,
        offerURL = req.offerURL,
        accessToken = prepareClaimResult.accessToken,
        offeredCredentialsProofRequests = prepareClaimResult.offeredCredentialsProofRequests,
        credentialIssuer = prepareClaimResult.resolvedCredentialOffer.credentialIssuer,
    )
    context.respond(HttpStatusCode.OK, responsePayload)
}
```

File: waltid-services/waltid-e2e-tests/src/test/kotlin/ExchangeExternalSignaturesApi.kt [hunk 5]
```diff
@@ -1,26 +1,37 @@
...
-import kotlinx.serialization.json.jsonPrimitive
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.decodeFromByteArray
+import kotlinx.serialization.json.*
 import kotlinx.uuid.UUID
-import kotlin.test.assertFalse
+import kotlin.test.assertNotEquals
+import kotlin.test.assertNotNull
 
 class ExchangeExternalSignatures {
```
Comment (mikeplotean) [inline waltid-services/waltid-e2e-tests/src/test/kotlin/ExchangeExternalSignaturesApi.kt:36]: not part of this pr, but just noticed
since this test suite wipes the wallet data before running the tests,
maybe it's worth mentioning that in e2e test, at this location:
```kotlin
// test External Signature API Endpoints
ExchangeExternalSignatures().executeTestCases()
```

File: waltid-services/waltid-e2e-tests/src/test/kotlin/ExchangeExternalSignaturesApi.kt [hunk 6]
```diff
@@ -168,21 +141,222 @@ class ExchangeExternalSignatures {
...
+                    did = if (useOptionalParameters) holderDID else null,
+                    offerURL = offerURL,
+                    credentialIssuer = prepareResponse.credentialIssuer,
+                    offeredCredentialProofsOfPossession = offeredCredentialProofsOfPossession,
+                    accessToken = prepareResponse.accessToken,
+                )
+            )
+        }.expectSuccess()
+        val credList = response.body<List<WalletCredential>>()
+        assert(credList.size == issuanceRequests.size) { "There should as many credentials in the wallet as requested" }
```
Comment (mikeplotean) [inline waltid-services/waltid-e2e-tests/src/test/kotlin/ExchangeExternalSignaturesApi.kt:346]: nitpicking alert!
perhaps typo for 'should be'?

**Review by mikeplotean — APPROVED** [review]

File: waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/usecase/claim/ClaimCommons.kt [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    )
+
+    fun addReceiveCredentialToUseCaseLog(
+        tenant: String,
+        account: UUID,
+        wallet: UUID,
+        credential: WalletCredential,
+        credentialType: String,
+        eventUseCase: EventLogUseCase,
+    ) {
```
[the remark this answers is not in the document]
Comment (cpatsonakis) [reply waltid-services/waltid-wallet-api/src/main/kotlin/id/walt/webwallet/usecase/claim/ClaimCommons.kt:36]: ok, will do

**Review by mikeplotean — APPROVED** [review]

**Comment by cpatsonakis** [comment]
Feedback from the conversations above will be incorporated during the September release.
